### ISSUE TITLE
in-memory write quit working at commit d62a9e623

### ISSUE THREAD

**Issue by gsjaardema** [Unidata/netcdf-c#1154]
> Please provide as much of the following information as you can, as applicable to the issue being reported.  Naturally, not all information is relevant to every issue, but the more information we have to start, the better!

### Environment Information

> Feel free to skip this if the issue is related to documentation, a feature request, or general discussion.

* What platform are you using? (please provide specific distribution/version in summary)
  * [ x] Linux
  * [ ] Windows
  * [ x] OSX
  * [ ] Other
  * [ ] NA
* 32 and/or 64 bit?
  * [ ] 32-bit
  * [ x] 64-bit
* What build system are you using?
  * [ x] autotools (`configure`)
  * [ x] cmake
* Can you provide a sample netCDF file or `C` code to recreate the issue?
  * [ ] Yes (please attach to this issue, thank you!)
  * [ ] No
  * [ x] Not at this time

### Summary of Issue
The in-memory write capability of NetCDF that my application was using broke at commit d62a9e623c7025f168d515cc9a3dd79d4a13dde7 which is " Fix the NC_INMEMORY code to work in all cases with HDF5 1.10."

I think that the problem occurs down in 
```
static int
memio_pad_length(ncio* nciop, off_t length)
{
    NCMEMIO* memio;
    size_t len = (size_t)length;
    if(nciop == NULL || nciop->pvt == NULL) return NC_EINVAL;
    memio = (NCMEMIO*)nciop->pvt;

    if(!memio->persist)
        return EPERM; /* attempt to write readonly file*/
    if(memio->locked)
	return NC_EDISKLESS;
```
The original file is created via a call to `nc_create()`.  It looks like `memio->persist` is not set, so the file is being treated as readonly.  This is due to the code:
```
    if(fIsSet(ioflags,NC_WRITE) && !fIsSet(ioflags,NC_NOCLOBBER) && memio->diskless)
	memio->persist = 1;
```
in `memio_new`.   Basically, `NC_WRITE` is not set.   

If I modify my code such that I call `nc_create` with the mode `NC_WRITE`, then the code seems to work correctly again.  

It looks like the code now requires that `nc_create` be called with the mode `NC_WRITE` if the in-memory capability is to be used.  I am fine modifying my code to add `NC_WRITE` to the mode when calling `nc_create()`, but I don't think that this is documented or should be required (Unless there is a reason I don't see for creating a file as readonly...)

**Comment by WardF** [comment]
Thank you very much @gsjaardema .  @DennisHeimbigner, your thoughts? Is this behavior that needs to be documented? Or should be changed?

**Comment by gsjaardema** [comment]
Here is the documentation from the current `inmemory.md`:
```
### Diskless File Open
Calling *nc_open()* using the mode flag *NC_DISKLESS* will cause
the file being opened to be read into memory. When calling *nc_close()*,
the file will optionally be re-written (aka "persisted") to disk. This
persist capability will be invoked if and only if *NC_WRITE* is specified
in the mode flags at the call to *nc_open()*.

### Diskless File Create
Calling *nc_create()* using the mode flag *NC_DISKLESS* will cause
the file to initially be created and kept in memory.
When calling *nc_close()*, the file will be written
to disk.
Note that if it is desired to create the file in memory,
but not write to a disk file, then one can either set
the NC_NOCLOBBER mode flag or one can call *nc_abort()*
instead of *nc_close()*.
```
The `NC_WRITE` is specified for `nc_open()`, but I don't see any reference to it for `nc_create()`
Here are also a sampling of the current examples:
```
examples/C/file.c:   if ((res = nc_create(FILENAME, NC_CLOBBER, &ncid)))
examples/C/filter_example.c:    CHECK(nc_create(TESTFILE, NC_NETCDF4|NC_CLOBBER, &ncid));
examples/C/format.c:       if ((res = nc_create(file, NC_CLOBBER, &ncid)))
examples/C/large_files.c:   stat = nc_create("rec2Gp.nc", NC_CLOBBER|NC_64BIT_OFFSET, &ncid);
examples/C/met4D.c:   if ((res = nc_create(FILENAME, NC_CLOBBER, &ncid)))
examples/C/parallel_vara.c:    err = nc_create_par(filename, cmode, MPI_COMM_WORLD, MPI_INFO_NULL, &ncid); FATAL_ERR
examples/C/pres_temp_4D_wr.c:   if ((retval = nc_create(FILE_NAME, NC_CLOBBER, &ncid)))
examples/C/quick_large_files.c:   if ((res = nc_create("example.nc", NC_CLOBBER|NC_64BIT_OFFSET, &ncid)))
examples/C/quick_small_files.c:   if ((res = nc_create("example.nc", NC_CLOBBER, &ncid)))
examples/C/sfc_pres_temp_more.c:   if ((retval = nc_create(FILE_NAME, NC_CLOBBER, &ncid)))
examples/C/sfc_pres_temp_wr.c:   if ((retval = nc_create(FILE_NAME, NC_CLOBBER, &ncid)))
examples/C/simple.c:   if ((res = nc_create("short.nc", NC_CLOBBER, &ncid)))
examples/C/simple_nc4_wr.c:   if ((retval = nc_create(FILE_NAME, NC_NETCDF4|NC_CLOBBER, &ncid)))
examples/C/simple_xy_nc4_wr.c:   if ((retval = nc_create(FILE_NAME, NC_NETCDF4, &ncid)))
examples/C/simple_xy_wr.c:   if ((retval = nc_create(FILE_NAME, NC_CLOBBER, &ncid)))
```
None of these are using the `NC_WRITE` for `nc_create()`.  Note that I am fine adding `NC_WRITE`, but just noting that this is a non-backward compatible change which should be documented more visibly.

**Comment by DennisHeimbigner** [comment]
No, in this case, I now remember doing this deliberately for nc_create.
I forgot that it was a change versus 4.6.1; by bad.
So the documentation should rule.

**Comment by DennisHeimbigner** [comment]
But after thinking about it some more, I think the old behavior (requiring
NC_WRITE) is actually the right thing to do because requiring the user to
use nc_abort() is more work that thould not be needed. It should be possible
to specify all the required behavior at the time of opening the file.

**Comment by DennisHeimbigner** [comment]
After more thought, I think the best flag to have used consistently would have been
NC_NOCLOBBER. But it is probably too late to do that.

**Comment by gsjaardema** [comment]
@DennisHeimbigner Can you point me to that documentation for `nc_create()` which specifies that `NC_WRITE` needs to be specified?  I can only see that for `nc_open()`, but I am probably overlooking something...

**Comment by DennisHeimbigner** [comment]
I see it in the file libdispach/dfile.c in release 4.6.1.
It is in the documentation preceding nc_create.
I see also however, that this version of the documentation
does not appear to be what one sees from the Unidata web site
which is dated March 15, 2018.
[Ward- why is this so out of date?]

**Comment by gsjaardema** [comment]
OK, I see that now.  It might be good to update `inmemory.md` which has conflicting information.

**Comment by gsjaardema** [comment]
I still think there is an issue with the current implementation.  In my code, I called `nc_create()`, and was performing output requests.  At some point, the NetCDF library called the `memio_pad_length` function which returned the `EPERM` error code stating that I was attempting to write to a readonly file.  This is the same file that I had been writing too previously. 

In other words, should the file be considered readonly just because I don't want it to persist?  I think the write calls to the file should be valid for a file created via `nc_create()` whether or not the file is going to be persisted.   

For example, I may want to create the in-memory representation of a new file and then access that memory through some other mechanism, but I don't want the file to persist.

Basically, I think the logic in `memio_pad_length` is incorrect for a file created via `nc_create()`.

There may be another use-case in which I open an existing file using `nc_open()' and I want it to be read into memory.  I then write to the file, but do not want it to persist.  In other words -- read/write access to an in memory representation of an existing file, but do not persist. 

Maybe an `NC_PERSIST` mode option would avoid a lot of confusion/complication.

**Comment by DennisHeimbigner** [comment]
I spotted this in my testing of the fix. The problem is that there are two notions
of write associated with a diskless file.
1. Modify the in-memory representation
2. Write the final result to disk.
The rule I am adopting is that NC_WRITE only refers to case 2.
This means that the in-memory file should always be modifiable,
but the presence/absence of NC_WRITE will determine if the file
is to  be pushed to disk.
I am not entirely happy with this and would like to implement the following cases.
1. in-memory is read-only and !persist
2. in-memory is read/write and !persist
3. in-memory is read/write and persist
4. in-memory is read-only and persist [ignored as meaningless]
If I want to get this set of cases, then it looks like I need use
two flags, say NC_WRITE to control if the in-memory is read/write or not
and NC_NOCLOBBER if the file is to be persisted.
Since you are one of the few users of diskless, your opinion is important.

**Comment by gsjaardema** [comment]
I prefer the addition of a new flag -- `NC_PERSIST`?  

I agree `NC_WRITE` should control whether an existing file can be written.  A file created by `nc_create()` should automatically have `NC_WRITE` applied in my opinion.

`NC_CLOBBER` should continue with current meaning -- if file being written to disk exists and `!NC_CLOBBER`, then failure whether in-memory or not. 

I guess my main view is that behavior should be very similar whether or not in-memory is used. If we continue to only use `NC_CLOBBER` and `NC_WRITE`, then I don't see how to get the current failure condition that I `nc_create()` a file and that file exists on disk, so should fail if `NC_CLOBBER` not specified.

**Comment by gsjaardema** [comment]
Note also that currently, `NC_WRITE` is not a valid flag if using PNetCDF as the dispatcher:
```
LEGAL_CREATE_FLAGS = (NC_NOCLOBBER | NC_64BIT_OFFSET | NC_CLASSIC_MODEL | NC_SHARE | NC_LOCK | NC_64BIT_DATA | NC_MPIIO);
 
LEGAL_OPEN_FLAGS = (NC_WRITE | NC_NOCLOBBER | NC_SHARE | NC_LOCK | NC_CLASSIC_MODEL | NC_64BIT_OFFSET | NC_64BIT_DATA | NC_MPIIO);
 ```

**Comment by DennisHeimbigner** [comment]
I agree and that is what I plan to implement: a new NC_PERSIST flag.

**Comment by DennisHeimbigner** [comment]
Update: I have a fix on the branch "disklesswrite.tmp". It must wait however
on a couple other pull requests.

**Comment by WardF** [comment]
@DennisHeimbigner Which branches are we waiting on? I can prioritize those.

**Comment by DennisHeimbigner** [comment]
let me verify; you may have already merged them and I did not notice.

**Comment by DennisHeimbigner** [comment]
Ok, at this point, the remaining pr I need to have integrated is https://github.com/Unidata/netcdf-c/pull/1147

**Comment by WardF** [comment]
#1147 is merged.

**Comment by DennisHeimbigner** [comment]
Fixed by https://github.com/Unidata/netcdf-c/pull/1158
